### PR TITLE
quincy: Implement CIDR blocklisting

### PR DESCRIPTION
**Pull request by gregsfortytwo**
backport tracker: https://tracker.ceph.com/issues/55746

---

backport of https://github.com/ceph/ceph/pull/44151
parent tracker: https://tracker.ceph.com/issues/53050

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

File: doc/rados/operations/control.rst [hunk 1]
```diff
@@ -243,18 +243,23 @@ Deployments utilizing Nautilus (or later revisions of Luminous and Mimic)
...
-Add/remove an IP address to/from the blocklist. When adding an address,
+Add/remove an IP address or CIDR range to/from the blocklist.
+When adding to the blocklist,
 you can specify how long it should be blocklisted in seconds; otherwise,
 it will default to 1 hour. A blocklisted address is prevented from
-connecting to any OSD. Blocklisting is most often used to prevent a
-lagging metadata server from making bad changes to data on the OSDs.
+connecting to any OSD. If you blocklist an IP or range containing an OSD, be aware
+that OSD will also be prevented from performing operations on its peers where it
+acts as a client. (This includes tiering and copy-from functionality.)
```
Comment (anthonyeleven) [inline doc/rados/operations/control.rst:252]: Should we explicitly mention backfill, recovery, replication?  Are those subsumed by "copy-from"?

**Review by anthonyeleven — APPROVED** [review]
Doc lgtm with one question about detail.

File: doc/rados/operations/control.rst (hunk 1 above)
Comment (gregsfortytwo) [reply doc/rados/operations/control.rst:252]: The OSD is not a client to other OSDs when doing backfill and other peering ops. We can update that but I'd rather do it against main and then backport the doc update.

**Comment by gregsfortytwo** [comment]
Jenkins test API

File: qa/workunits/cephtool/test.sh [hunk 2]
```diff
@@ -1426,11 +1426,20 @@ function test_mon_osd()
...
+  expect_false "ceph osd blocklist add 1234.56.78.90/100"
+
+  # test range blocklisting
+  bl=192.168.0.1:0/24
+  ceph osd blocklist range add $bl
+  ceph osd blocklist ls | grep $bl
+  ceph osd blocklist range rm $bl
+  ceph osd blocklist ls | expect_false grep $bl
+  bad_bl=192.168.0.1/33
+  expect_false ceph osd blocklist range add $bad_bl
```
Comment (ljflores) [inline qa/workunits/cephtool/test.sh:1442]: @gregsfortytwo I'm seeing this in the rados suite:

http://pulpito.front.sepia.ceph.com/yuriw-2022-06-01_02:28:14-rados-wip-yuri-testing-2022-05-31-1642-quincy-distro-default-smithi/6856753/
```
2022-06-01T03:34:44.256 INFO:tasks.workunit.client.0.smithi066.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1437: test_mon_osd:  ceph osd blocklist range add 192.168.0.1/24
2022-06-01T03:34:45.650 INFO:tasks.workunit.client.0.smithi066.stderr:blocklisting cidr:192.168.0.1:0/24 until 2022-06-01T04:34:44.774003+0000 (3600 sec)
2022-06-01T03:34:45.663 INFO:tasks.workunit.client.0.smithi066.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1438: test_mon_osd:  ceph osd blocklist ls
2022-06-01T03:34:45.663 INFO:tasks.workunit.client.0.smithi066.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1438: test_mon_osd:  grep 192.168.0.1/24
2022-06-01T03:34:46.112 INFO:tasks.workunit.client.0.smithi066.stderr:listed 7 entries
2022-06-01T03:34:46.122 INFO:tasks.workunit.client.0.smithi066.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_mon_osd:  rm -fr /tmp/cephtool.kqr
2022-06-01T03:34:46.124 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-06-01T03:34:46.124 INFO:tasks.workunit:Stopping ['cephtool'] on client.0...
2022-06-01T03:34:46.125 DEBUG:teuthology.orchestra.run.smithi066:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2022-06-01T03:34:46.380 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
```
Has this been fixed in the new commits you added?

**Comment by ljflores** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2022-06-01_02:28:14-rados-wip-yuri-testing-2022-05-31-1642-quincy-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/yuriw-2022-06-01_14:00:31-rados-wip-yuri-testing-2022-05-31-1642-quincy-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/yuriw-2022-06-02_20:23:28-rados-wip-yuri-testing-2022-06-02-0810-quincy-distro-default-smithi/

Failures, unrelated:
    1. https://tracker.ceph.com/issues/52321
    2. https://tracker.ceph.com/issues/49287
    3. https://tracker.ceph.com/issues/52652
    4. https://tracker.ceph.com/issues/53575
    5. https://tracker.ceph.com/issues/55741

Details:
    1. qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator - Ceph - Orchestrator
    2. podman: setting cgroup config for procHooks process caused: Unit libpod-$hash.scope not found - Ceph - Orchestrator
    3. ERROR: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest) - Ceph - Mgr
    4. Valgrind reports memory "Leak_PossiblyLost" errors concerning lib64 - Ceph - RADOS
    5. cephadm/test_dashboard_e2e.sh: Unable to find element `cd-modal .custom-control-label` when testing on orchestrator/01-hosts.e2e-spec.ts - Ceph - Mgr - Dashboard

File: qa/workunits/cephtool/test.sh (hunk 2 above)
Comment (ljflores) [reply qa/workunits/cephtool/test.sh:1442]: This does contain the cephtool fix. I checked a subsequent rerun after the branch was rebased, and this problem does not appear anymore.

**Review by vshankar — APPROVED** [review]

**Pull request merged**